### PR TITLE
use `on_unimplemented` for better error messages

### LINKED ISSUE #2: [suggestion] include `build` method when all required fields are not available.
In order for users to know which fields are missing we can do something like this:

```rust
struct Builder {
   req1: Option<usize>,
   req2: Option<usize>,
   opt1: usize
}

impl Builder<false, false> {
    fn build(&self, req1: usize, req2: usize) -> ...
}

impl Builder<false, true> {
    fn build(&self, req1: usize) -> ...
}

impl Builder<true, false> {
    fn build(&self, req2: usize) -> ...
}

impl Builder<true, true> {
   fn build(&self) -> ...
}

```

The only problem is that it generates a lot of code and it may not be ideal. But if you are OK with the design, I can send a PR if you want to.

### PR DESCRIPTION
**Pull request by sahandevs**
closes #2 

I don't have an easy way to write a UI test for this. Any ideas?

image

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -399,7 +395,23 @@ pub fn builder(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
...
+                            #[rustc_on_unimplemented(
+                                message=#message,
+                                label=#label,
+                            )]
+                        });
+                    }
+
                     constraint_traits.push(quote! {
+                        #(#error_message)*
                         trait #trait_name {}
```
Comment (maminrayej) [inline src/lib.rs:415]: Shouldn't we have a single message per `constraint_trait`? Why `error_message` is a `vec`?

File: Cargo.toml [hunk 2]
```diff
@@ -9,6 +9,9 @@ license = "MIT"
 categories = ["rust-patterns"]
 keywords = ["builder", "builder_pattern"]
 
+[features]
+better_error_message = []
```
Comment (maminrayej) [inline Cargo.toml:13]: May I suggest a more concise name?
How about `better_error`? It kinda rhythms :)

**Comment by maminrayej** [comment]
I didn't understand your problem with writing a UI test unfortunately. Can you elaborate?

**Comment by sahandevs** [comment]
Ok! I've sold my soul to the devil to add some tests. :))
also, I've addressed the issues that you mentioned. would you please review it again? @maminrayej

**Comment by maminrayej** [comment]
Wow you really went above and beyond to make it work huh? :))
All is well and I'll merge this as is, but I want to discuss the matter here a bit for future reference.

It's my understanding that not being able to tell `cargo` which feature should be enabled for which test is our main problem right?
Also we can't run some tests using stable Rust and some other using nightly(using one `cargo` command).

**Comment by sahandevs** [comment]
Exactly, we can't specify crate features and compiler targets using the `try_build` crate.

**Comment by maminrayej** [comment]
@sahandevs Could you look at this [branch](https://github.com/maminrayej/tidy-builder/tree/split-tests) as an alternative?
It uses the names of the test files to distinguish stable tests from nightly tests, and uses a `Makefile` to make running the tests easier. It's still a hacky way to do things, but it extracts the logic of splitting tests from the tests themselves (the `ui.rs` file does not need to spawn a new `cargo test`) and delegates it to a `Makefile`.

**Pull request merged**